concepts/recursion/about.md

```diff
@@ -0,0 +1,98 @@
+# About
+
+Recursion is a way to repeat code in a function by the function calling itself.
+It is another way to loop/iterate. Like looping, the function should test a Boolean expression to know when to end.
+Unlike looping, recursion that never ends will not run infinitely.
+The values used in each function call will be placed in their own frame on the stack.
+If the amount of function calls takes up more space than the stack has, it will result in a stack overflow error.
+
+```python
+def printIncrement(i, maxValue):
+    if i > maxValue:
+        return
+    print(f'The number is {i}')
+    printIncrement(i + 1, maxValue)
+
```

Comment:
```suggestion
```python
def print_increment(step, max_value):
    if step > max_value:
        return
    print(f'The number is {step}')
    print_increment(step + 1, max_value)

```

Reply:
So this is the obligatory "PEP8 and examples need to be exemplars" nag.  Sorry I need to do this.  The gist is:
1.  snake_case function names and variable names.
2.  no single letter variable names - not even in comprehensions.
3.  REPL formatting for examples, if at all possible (example below).

```python
>>> def print_increment(step, max_value):
...     if step > max_value:
...         return
...     print(f'The number is {step}')
...     print_increment(step + 1, max_value)

```

Not sure here and below how to make that work, exactly.  We may want to replace this section with diagrams on the same function looping vs recursion, or with a link to Python Tutor like [this one](https://pythontutor.com/visualize.html#code=def%20steps_recurse%28number,%20step%3D0%29%3A%0A%20%20%20%20%22%22%22Calculate%20number%20of%20steps%20to%20reach%201.%0A%20%20%20%20%0A%20%20%20%20%3Aparam%20number%3A%20int%20-%20number%20to%20run%20through%20the%20conjecture.%0A%20%20%20%20%3Aparam%20step%3A%20int%20-%20current%20step%20number/iteration.%0A%20%20%20%20%3Areturn%3A%20step%3A%20int%20-%20total%20number%20of%20steps%20taken%20to%20reach%201.%0A%20%20%20%20%22%22%22%0A%0A%20%20%20%20new_number%20%3D%20number//2%20if%20number%20%25%202%20%3D%3D%200%20else%20%28number%20*%203%29%20%2B%201%0A%20%20%20%20%0A%20%20%20%20return%20step%20if%20number%20%3D%3D%201%20else%20steps_recurse%28new_number,%20step%3Dstep%20%2B%201%29%0A%0A%0A%0Adef%20steps_loop%28number%29%3A%0A%20%20%20%20%22%22%22Calculate%20number%20of%20steps%20to%20reach%201.%0A%20%20%20%20%0A%20%20%20%20%3Aparam%20number%3A%20int%20-%20number%20to%20run%20through%20the%20conjecture.%0A%20%20%20%20%3Aparam%20step%3A%20int%20-%20current%20step%20number/iteration.%0A%20%20%20%20%3Areturn%3A%20step%3A%20int%20-%20total%20number%20of%20steps%20taken%20to%20reach%201.%0A%20%20%20%20%22%22%22%0A%0A%20%20%20%20step%20%3D%200%20%0A%20%20%20%20while%20number%20!%3D%201%3A%0A%20%20%20%20%20%20%20%20number%20%3D%20number//2%20if%20number%20%25%202%20%3D%3D%200%20else%20%28number%20*%203%29%20%2B%201%0A%20%20%20%20%20%20%20%20step%20%2B%3D%201%0A%20%20%20%20%0A%20%20%20%20return%20step%0A%0Asteps_recurse%285%29%0Asteps_loop%285%29%0A%0A&cumulative=false&heapPrimitives=nevernest&mode=edit&origin=opt-frontend.js&py=3&rawInputLstJSON=%5B%5D&textReferences=false).

Reply:
Thanks! I ran the code through [pep8online](http://pep8online.com/) until there were no issues reported, but that was evidently not sufficient. I'll probably take Kotlin off my machine and put Python back on with a good linter.